expand.go

```diff
@@ -0,0 +1,213 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"bytes"
+	"io"
+
+	"golang.org/x/text/transform"
+)
+
+// expandReadCloser implements io.ReadCloser
+type expandReadCloser struct {
+	io.Reader
+	io.Closer
+}
+
+func newExpandReadCloser(rc io.ReadCloser, expandFunc func(string) (string, error)) io.ReadCloser {
+	return &expandReadCloser{
+		transform.NewReader(
+			rc,
+			&expandTransformer{expand: expandFunc}),
+		rc,
+	}
+}
+
+// expandTransformer implements transform.Transformer
+type expandTransformer struct {
+	transform.NopResetter
+	expand func(string) (string, error)
+}
+
+// First char of shell variable may be [a-zA-Z_]
+func isShellNameFirstChar(c byte) bool {
+	return c == '_' ||
+		(c >= 'a' && c <= 'z') ||
+		(c >= 'A' && c <= 'Z')
+}
+
+// Char's after the first of shell variable may be [a-zA-Z0-9_]
+func isShellNameChar(c byte) bool {
+	return c == '_' ||
+		(c >= 'a' && c <= 'z') ||
+		(c >= 'A' && c <= 'Z') ||
+		(c >= '0' && c <= '9')
+}
+
+// bytesIndexCFunc returns the index of the byte for which the
+// complement of the supplied function is true
+func bytesIndexCFunc(buf []byte, f func(b byte) bool) int {
+	for i, b := range buf {
+		if !f(b) {
+			return i
+		}
+	}
+	return -1
+}
+
+// Transform expands shell-like sequences like $foo and ${foo} using
+// the configured expand function.  The sequence '$$' is replaced with
+// a literal '$'.
+func (e *expandTransformer) Transform(dst, src []byte, atEOF bool) (int, int, error) {
+	var srcPos int
+	var dstPos int
+
+	for srcPos < len(src) {
+
+		if dstPos == len(dst) {
+			return dstPos, srcPos, transform.ErrShortDst
+		}
+
+		end := bytes.IndexByte(src[srcPos:], '$')
+
+		if end == -1 {
+			// src does not contain '$', copy into dst
+			cnt := copy(dst[dstPos:], src[srcPos:])
+			srcPos += cnt
+			dstPos += cnt
+			continue
+		} else if end > 0 {
+			// copy chars preceding '$' from src to dst
+			cnt := copy(dst[dstPos:], src[srcPos:srcPos+end])
+			srcPos += cnt
+			dstPos += cnt
+
+			if dstPos == len(dst) {
+				return dstPos, srcPos, transform.ErrShortDst
+			}
+		}
+
+		// src[srcPos] now points to '$', dstPos < len(dst)
+
+		// If we're at the end of src, but we found a starting
+		// token, return ErrShortSrc, unless we're also at EOF,
+		// in which case just copy it dst.
+		if srcPos+1 == len(src) {
+			if atEOF {
+				dst[dstPos] = src[srcPos]
+				srcPos++
+				dstPos++
+				continue
+			}
+			return dstPos, srcPos, transform.ErrShortSrc
+		}
+
+		// At this point we know that src[srcPos+1] is populated.
+
+		// If this token sequence represents the special '$$'
+		// sequence, emit a '$' into dst.
+		if src[srcPos+1] == '$' {
+			dst[dstPos] = src[srcPos]
+			srcPos += 2
+			dstPos++
+			continue
+		}
+
+		var token []byte
+		var tokenEnd int
```

Comment:
nit: var groups are cool

```go
var (
  token []byte
  tokenEnd int
)
```

Reply:
Oh let me count the multitude of ways by which a variable can be defined.  I think Rob Pike forgot that your design isn't done when there's nothing left to add, it's done when there's nothing left to take away :-)